### PR TITLE
[torch/multiprocessing] Add start_process_in_context method that takes a multiprocessing context from user

### PR DESCRIPTION
**Pull request by kiukchung**
Summary: Allows user to pass a multiprocessing context so that they can create any multiprocessing datastructures on them to use in their sub-processes.

Test Plan: sandcastle

Differential Revision: D20104741

### CODE REVIEW

**Review by agolynski — APPROVED** [review]
LGTM

**Review by ailzhang — CHANGES_REQUESTED** [review]
Hi @kiukchung, thanks for the patch! 
Can you point me to the usecase of this API? Also please add a test in `test_multiprocessing_spawn.py`.

**Comment by kiukchung** [comment]
@ailzhang  take a look at https://github.com/pytorch/elastic/blob/master/test/timer/local_timer_example.py#L52

TLDR; its useful when you want to use a multiprocessing datastructure to share data between the processes. In the example linked above, I've implemented an expiration timer where an arbitrary code-block in the subprocess can be wrapped with an expiration time. If the code-block does not finish within the specified expiry the main process will `os.kill` the subprocess. A multiprocessing Queue is used to pass acquire and release messages between the main process and subprocesses and this queue has to be created in the same context as the subprocess, which is currently not possible in the apis provided by `torch.multiprocessing`. 

Will add usage example in the docs and add a test.

**Comment by ailzhang** [comment]
@kiukchung  Please correct me if I'm wrong, `mp_context` is very similar to `start_method`, maybe we can just add this keyword arg to the existing `start_processes` API and properly handle them? 
I'm a little worried about we have too many APIs, each doing a little bit different. In that case, having `start_processes` being a generic API might be preferred.

**Comment by kiukchung** [comment]
@ailzhang  is your suggestion overloading the type of `start_method` in `start_processes()`? It could either be a string ("spawn", "fork", etc) or a multiprocessing context. For instance one could call:
```
start_processes(fn, args=(...), nprocs=2, start_method="spawn")
# -- or --
ctx = multiprocessing.get_context("spawn")
start_processes(fn, args=(...), nprocs=2, start_method=ctx)
```
I thought about this but concluded that this was more confusing than having an explicit method that makes the intention clear.

I also thought about adding a `context` parameter to `start_processes` so something like

```
def start_processes(fn, args=(), nprocs=1, join=True, daemon=False, start_method='spawn', context=None):
```

however this leads to some pretty nasty parameter checking (only one of start_method or context can be specified but since `start_method` defaults to `spawn`, to launch the processes in a context, the user has to call `start_processes(fn, args, nprocs, start_method=None, context=ctx)`

**Comment by kiukchung** [comment]
@ailzhang thanks for the offline discussion so the following worked:

```
import multiprocessing as mp
start_method="spawn" # or fork or forkserver
q = mp.get_context(start_method).Queue()
torch.multiprocessing.start_processes(fn, args=(q,), nprocs=4, start_method)
```

Looks like python multiprocessing returns the same context object if the same start_method is given:

```
c1 = mp.get_context("spawn")
c2 = mp.get_context("spawn")
c1 == c2 (is true)
```
